[ZIP 244] Changes to transparent component of signature digest (#577)
zcash/zips: zip-0244.rst

S.2: transparent_sig_digest
```````````````````````````
In the case that transparent inputs or outputs are present, the value of
``transparent_sig_digest`` depends upon the value of a ``hash_type`` flag.

The construction of each component below depends upon the values of the
``hash_type`` flag bits. Each component will be described separately.

This digest is a BLAKE2b-256 hash of the following values ::

   S.2a: hash_type               (1 byte)
   S.2b: prevouts_sig_digest     (32-byte hash)
   S.2c: amounts_sig_digest      (32-byte hash)
   S.2d: script_codes_sig_digest (32-byte hash)
   S.2e: sequence_sig_digest     (32-byte hash)
   S.2f: outputs_sig_digest      (32-byte hash)
   S.2g: txin_sig_digest         (32-byte hash)

The personalization field of this hash is set to::

   "ZTxIdTranspaHash"

> In the case that the transaction has no transparent components,

Reviewer: ```suggestion
In the case that the transaction has no transparent inputs or outputs,
```

``transparent_sig_digest`` is ::

  BLAKE2b-256("ZTxIdTranspaHash", [])

S.2a: hash_type
'''''''''''''''
This is an 8-bit unsigned value. The ``SIGHASH`` encodings from the legacy

> script system are reused: one of ``SIGHASH_ALL``, ``SIGHASH_NONE``, and
> ``SIGHASH_SINGLE``, with or without the ``SIGHASH_ANYONECANPAY`` flag. The
> following restrictions apply, which cause validation failure if violated:

Reviewer 2: ```suggestion
script system are reused: one of ``SIGHASH_ALL`` (0x01), ``SIGHASH_NONE`` (0x02),
and ``SIGHASH_SINGLE`` (0x03), with or without the ``SIGHASH_ANYONECANPAY``
flag (0x80). The following restrictions apply, which cause validation failure if violated:
```

- Using any undefined ``hash_type`` (not 0x01, 0x02, 0x03, 0x81, 0x82, or 0x83).
- Using ``SIGHASH_SINGLE`` without a "corresponding output" (an output with the
  same index as the input being verified).

If we are producing a hash for the signature over a Sapling Spend or an Orchard

> Action, ``hash_type`` is set to ``SIGHASH_ALL``, and is not encoded anywhere in

Reviewer: I find this confusing. What does "is set to" pertain to?

Author: I don't know how to write this in any plainer English: `hash_type` is set to `SIGHASH_ALL`. That is, this 1-byte input to the transparent sig digest is set to a constant, rather than taken from the transaction encoding.

Reviewer: Maybe an addendum: "hash_type is set to SIGHASH_ALL and the the hashes `S.2b..S.2g` are computed in accordance with this flag value."

I think it's the "is not encoded anywhere" that threw me off - it's not encoded, but it is committed to in `S.2a`?

Reviewer 3: Resolved by suggested changes commit

This digest is a BLAKE2b-256 hash of the following values ::

   S.2a: hash_type               (1 byte)
   S.2b: prevouts_sig_digest     (32-byte hash)
   S.2c: amounts_sig_digest      (32-byte hash)
   S.2d: script_codes_sig_digest (32-byte hash)
   S.2e: sequence_sig_digest     (32-byte hash)
   S.2f: outputs_sig_digest      (32-byte hash)
   S.2g: txin_sig_digest         (32-byte hash)

The personalization field of this hash is set to::

   "ZTxIdTranspaHash"

In the case that the transaction has no transparent inputs or outputs,
``transparent_sig_digest`` is ::

  BLAKE2b-256("ZTxIdTranspaHash", [])

S.2a: hash_type
'''''''''''''''
This is an 8-bit unsigned value. The ``SIGHASH`` encodings from the legacy
script system are reused: one of ``SIGHASH_ALL`` (0x01), ``SIGHASH_NONE`` (0x02),
and ``SIGHASH_SINGLE`` (0x03), with or without the ``SIGHASH_ANYONECANPAY`` flag
(0x80). The following restrictions apply, which cause validation failure if
violated:

- Using any undefined ``hash_type`` (not 0x01, 0x02, 0x03, 0x81, 0x82, or 0x83).
- Using ``SIGHASH_SINGLE`` without a "corresponding output" (an output with the
  same index as the input being verified).

> If we are producing a hash for the signature over a Sapling Spend or an Orchard
> Action, ``hash_type`` is set to ``SIGHASH_ALL``, and is not encoded anywhere in
> the transaction's authorizing data.

Author: ```suggestion
If we are producing a hash for the signature over a transparent input, the value
of ``hash_type`` is obtained from the input's ``scriptSig`` as encoded in the
transaction. If we are producing a hash for the signature over a Sapling Spend
or an Orchard Action, ``hash_type`` is set to ``SIGHASH_ALL``.
```

   BLAKE2b-256(``ZTxIdPrevoutHash``, [])

S.2c: amounts_sig_digest
''''''''''''''''''''''''
If the ``SIGHASH_ANYONECANPAY`` flag is not set, the value of
``amounts_sig_digest`` is a BLAKE2b-256 hash of the concatenation of the 8-byte
signed little-endian representations of all ``value`` fields [#bdr-txout]_ for
the coins spent by the transparent inputs to the transaction.

The personalization field of this hash is set to::

  "ZTxTrAmountsHash"

If the ``SIGHASH_ANYONECANPAY`` flag is set, ``amounts_sig_digest`` is::

  BLAKE2b-256("ZTxTrAmountsHash", [])

S.2d: script_codes_sig_digest
'''''''''''''''''''''''''''''
If the ``SIGHASH_ANYONECANPAY`` flag is not set, the value of
``script_codes_sig_digest`` is a BLAKE2b-256 hash of the concatenation of the

> field encodings (including leading ``CompactSize``) of all ``pk_script`` fields
> [#bdr-txout]_ for the coins spent by the transparent inputs to the transaction.

Reviewer 2: ```suggestion
field encodings (each including a leading ``CompactSize``) of all ``pk_script``
fields [#bdr-txout]_ for the coins spent by the transparent inputs to the
transaction.
```

   BLAKE2b-256(``ZTxIdOutputsHash``, [])

S.2g: txin_sig_digest
'''''''''''''''''''''
If we are producing a hash for the signature over a transparent input, the value
of ``txin_sig_digest`` is a BLAKE2b-256 hash of the following properties of the
transparent input being signed, initialized with the personalization field value
``Zcash___TxInHash`` (3 underscores)::

   S.2g.i:   prevout     (field encoding)
   S.2g.ii:  value       (8-byte signed little-endian)

>    S.2g.iii: script_code (field encoding)

Reviewer 4: I am wondering whether there is a difference between `script_code`, `pk_script` (used in S.2d) and `scriptPubKey` (used in T.2c). In Bitcoin `script_code` and `scriptPubKey` have different meanings, one reason is SegWit BIP141 and the other is due to `OP_CODESEPARATOR`, but I think neither of these applies in Zcash. If they all have the same meaning, then why not call them all `scriptPubKey` in this spec? And similarly rename `script_codes_sig_digest` to `scriptpubkeys_sig_digest`, so that the naming is more similar to BIP341.

Author: All three names refer to the same field in `CTxOut`, and mean exactly the same thing: the transparent Bitcoin Script that the transaction output commits to. I've moved to calling it `scriptCode` (as Bitcoin did when they added SegWit), considering `scriptPubKey` to be the legacy inaccurate name (since it rarely if ever directly contained a public key in Bitcoin at the time Zcash forked, and in Zcash is almost exclusively either P2PKH or P2SH, containing only hash commitments). You're right that the ZIP shouldn't have both `scriptPubKey` and `scriptCode`; this should be fixed.

I don't know for certain why BIP 341 chose to revert from `scriptCode` back to `scriptPubKey`, but I suspect it's because Taproot is explicitly a combination of a public key and the hash of a script, so the field contains a public key directly again:
> - Let q be the 32-byte array containing the witness program (the second push in the scriptPubKey) which represents a public key according to BIP340.

That rationale doesn't apply to Zcash, but even if we were deploying Taproot, I'd still prefer `scriptCode` over `scriptPubKey`.

`pk_script` is only mentioned because we're referencing the Bitcoin Developer Reference for the definition of `CTxOut`, which uses that name.

Reviewer: https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki#cite_note-17, which we cite, refers to this as `scriptPubKey`, but we can refer to it as `scriptCode` internally and just keep the reference. Also, in looking at this I noticed that there's an off-by-one error in our cite_note links; I'm not sure how this happened but I'll fix it.

Reviewer 4: > All three names refer to the same field in `CTxOut`, and mean exactly the same thing: the transparent Bitcoin Script that the transaction output commits to.

In than case, what you want is `scriptPubKey`.

> I've moved to calling it `scriptCode` (as Bitcoin did when they added SegWit)

In Bitcoin `scriptCode` and `scriptPubKey` are two different things and Bitcoin never stopped using the term `scriptPubKey`, certainly not with SegWit, see BIP-143.

In addition to the two differences that I mentioned in my original comment, there is one more which is easy to illustrate. Consider a P2SH which consists of a 2-of-3 multisig. The `scriptPubKey` and `scriptCode` will look like this:
`scriptPubKey`: `A9 14 <20-byte-scripthash> 87`
`scriptCode`: `52 21 <pubkey1> 21 <pubkey2> 21 <pubkey3> 53 AE`

> I don't know for certain why BIP 341 chose to revert from `scriptCode` back to `scriptPubKey`

That never happened. BIP 341 ceased to use `scriptCode` as such.

> even if we were deploying Taproot, I'd still prefer `scriptCode` over `scriptPubKey`.

The established terminology might not be perfect, but mixing it up will not help anyone.

Reviewer 2: I believe @andrewkozlik is correct, and I'm inclined to revert those parts of 76433bbdb452467aa70fbc2df12a76e4cc958f60. We need to check whether there are other incorrect usages.

Author: We've just had a meeting to figure out what was going on here, where my understanding went wrong, and where the above comments were correct or not.

---

BIP 143 defined `scriptCode` as being one of two things:
- For P2WPKH, the equivalent of a `scriptPubKey` for P2PKH (the opcodes are AFAICT identical).
- For P2WSH, the equivalent of a `redeemScript` extracted from `scriptSig` for P2SH (if you look at the non-`OP_CODESEPARATOR` branch).

and did not alter non-SegWit sighashes (i.e. P2PKH or P2SH).

BIP 341 defines two sets of script-related components:
- At a transaction level, the `scriptPubKey` field for every input is digested (`sha_scriptpubkeys`).
- At an input level, the `scriptPubKey` field of that input.

> The wording Bitcoin chose to use in their BIPs was incidentally very confusing, using "serialized as" (metaphor) instead of "serialized like" (simile) or the more precise "seralized using the encoding of":
> - BIP 143: "`scriptCode` of the input (serialized as scripts inside `CTxOut`s)"
>   - Refers only to the format of the encoding.
> - BIP 341:
>   - "the SHA256 of all spent outputs' `scriptPubKeys`, serialized as script inside `CTxOut`."
>   - "`scriptPubKey` of the previous output spent by this input, serialized as script inside `CTxOut`."
>     - Refers to the format of the encoding, but is also valid to interpret as the actual encoded field values in the outputs.

---

In ZIP 143 (and maintained in ZIP 243), we mapped the SegWit P2WPKH and P2WSH semantics to P2PKH and P2SH, and defined `scriptCode` as:
- `redeemScript` if P2SH.
- `scriptPubKey` in all other cases.

This commits to the same data as if it just used `scriptPubKey` in all cases, because the way Bitcoin implemented P2SH, the consensus code only reaches that point if the `scriptPubKey` has a very specific format in which everything about it ends up being checked prior to using `redeemScript`. I believe at the time we kept the SegWit formulation because we wanted to stick close to the SegWit semantics, but in hindsight it would have been simpler to just use `scriptPubKey` everywhere.

Meanwhile, in this PR to ZIP 244 we are attempting to map BIP 341's Taproot semantics to P2PKH and P2SH. If we were to follow them closely, we would need to use `scriptPubKey` in the digests at both the transaction level (for all inputs) and the input level (per-input). However, in this PR (and in ZIP 244 prior to this PR), at the input level we continue to use `scriptCode` (i.e. for P2SH we use `redeemScript` instead of `scriptPubKey` in its digest).

---

So, conclusions:
- I was partially incorrect to say that `scriptPubKey`, `scriptCode`, and `pk_script` were all the same in this context. They _are_ all identical for P2PKH (and in fact for _every_ script except P2SH), while for P2SH `scriptPubKey` and `pk_script` are identical but `scriptCode` is not.
- We should rename `script_codes_sig_digest` to `scriptpubkeys_sig_digest`, to match BIP 341.
- [Consensus change]: We change the per-input data to commit to `scriptPubKey` instead of `scriptCode`
  - For P2PKH this is a no-op.
  - For P2SH, this means committing to the Hash160 digest of the `redeemScript` instead of the actual `redeemScript`. We discussed this and believe it is fine, because:
    - An adversary trying to take a valid signature and use it with a different `redeemScript` would need to perform a preimage attack on the Hash160 digest, which doesn't have a square-root optimization anywhere.
    - Even if they _could_ find a preimage, all that allows them to do is alter the input's `scriptSig` field; this doesn't alter the effecting data of the transaction, which by definition means the transaction has the same effect under consensus (spends the same inputs and produces the same outputs).

The block header byte format and version are not altered by this ZIP.

=========
Rationale
=========

In S.2, we use the same personalization strings for fields that have matching
fields in T.2, in order to facilitate reuse of their digests. In particular, the
"no transparent inputs or outputs" case of S.2 is identical to the equivalent

> case in T.2; thus for purely-shielded transactions, ``signature_digest`` is

Reviewer 2: ```suggestion
case in T.2; thus for fully shielded transactions, ``signature_digest`` is
```

- The semantics of ``outputs_sig_digest`` were changed, via a new consensus rule
  that rejects transparent inputs for which ``SIGHASH_SINGLE`` is set without a
  corresponding transparent output at the same index. BIP 341 does not give a
  rationale for this change, but without it these inputs were effectively using
  ``SIGHASH_NONE``, which is silently misleading.
- The semantics of ``txin_sig_digest`` were changed, to always commit to the
  ``scriptPubKey`` field of the transparent coin being spent, instead of the
  script actually being executed at the time ``signature_digest`` is calculated.

  - This ensures that the signature commits to the entire committed script. In
    Taproot, this makes it possible to prove to a hardware wallet what (unused)
    execution paths exist [#bip-0341-scriptPubKey]_. Alternate execution paths
    don't exist for P2PKH (where the executed script is ``scriptPubKey``) or
    P2SH (where ``scriptPubKey`` is fully executed prior to ``redeemScript``).
  - For P2SH, this means we commit to the Hash160 digest of ``redeemScript``
    instead of the actual script. Note that the Bitcoin P2SH design depends

>     entirely on Hash160 being preimage resistant, because otherwise anyone would

Reviewer 2: ```suggestion
    entirely on Hash160 being preimage-resistant, because otherwise anyone would
```

    be able to spend someone else's P2SH UTXO using a preimage. We do need to
    ensure that there is no collision attack; this holds because even if an

>     adversary could find a Hash160 collision, this would only enable them to

Reviewer 2: ```suggestion
    adversary could find a Hash160 collision, it would only enable them to
```
